[vernac] Remove VernacGoal, allow anonymous definitions in VernacDefinition

vbgl opened the pull request:
This is a preliminary steps for the implementation of attributes.

maximedenes commented on intf/vernacexpr.ml:
```diff
@@ -167,6 +171,11 @@ type option_ref_value =
+
+let name_of_ident_decl (id: ident_decl) : name_decl =
+  let (i, u) = id in
+  lname_of_lident i, u
```
Right, it may be safer to move this somewhere else.

ppedrot replied:
Yeah, we should definitely move back the interface-only files to mli.

ejgallego replied:
> Yeah, we should definitely move back the interface-only files to mli.

Please let's avoid introducing mli-only files again [`mli+ml`] is fine. For reasons see the `dune` readme.

Zimmi48 replied:
Dune is the ex-jbuilder FTR.

vbgl commented:
Thanks for your comments. I’ve moved the function definitions out of `intf` into `parsing/g_vernac.ml4` where there are used.

Following Gaëtan’s remark, `VernacStartTheoremProof` is much like `Fixpoint`: it is for mutual interactive definitions. Therefore, in a second commit, I’ve adjusted the type of this constructor to make explicit that such definitions always have names.

SkySkimmer commented:
>Following Gaëtan’s remark, VernacStartTheoremProof is much like Fixpoint: it is for mutual interactive definitions.

`Lemma foo : bar.` is a `VernacStartTheoremProof`, and except for the name it's like a `Goal bar.`

In fact why do we even have the `ProveBody` instead of using `VernacStartTheoremProof`?

ejgallego commented on parsing/g_vernac.ml4:
```diff
@@ -133,6 +133,14 @@ let test_plural_form_types loc kwd = function
+let lname_of_lident (i: lident) : lname =
+    let (loc, s) = i in
+    loc, Name s
+
```
Maybe `Loc.map` should be used here?

ejgallego commented on parsing/g_vernac.ml4:
```diff
@@ -133,6 +133,14 @@ let test_plural_form_types loc kwd = function
+let name_of_ident_decl (id: ident_decl) : name_decl =
+  let (i, u) = id in
+  lname_of_lident i, u
+
```
I think `on_fst` could be used here.

maximedenes commented:
> in particular I dislike the assert false.

I believe the `assert false` reflects a defect of today's system. I don't think there's any reason why requiring a name is correlated to providing no body. In particular, `Program` doesn't support `Goal`, which it probably should.

vbgl commented:
Here is a summary of (my view of) the current situation.
There are five constructors of `vernac_expr` which correspond to the beginning of a definition.
They vary in how many objects are defined, whether they have names (as specified by the user), and whether their body is given:

 -  Definition: one, named, optional body
 -  StartTheoremProof: many, named, no body
 -  Fixpoint: many, named, optional bodies
 -  Instance: one, optional name, optional body
 -  Goal: one, no name, no body

There is some overlap between all these cases
(for instance, `StartTheoremProof` seems a special case of `Fixpoint`).

I’d like to be able to see `Instance` as an attribute that could decorate some of the other constructs.

IMHA, an instance definition is a definition decorated with some annotations (this is an instance, here is its priority, etc.);
moreover, both constructors for `Instance` and `Definition` are (syntactically) very close.
This is why I try to merge them.
Instances can be anonymous, this is what motivated this PR:
allow definitions to have no explicit name.

As a side effect, the `Goal` constructor becomes a special case of an anonymous definition without body; hence the removal.

An other possibility could be to give names to everything very early, e.g., at parsing time. One issue I foresee is that the current naming scheme for anonymous instances needs to resolve the class (as the name of the class appears in the generated name for the instance); which might not be feasible at parsing time.

ejgallego commented:
Interesting, thanks for the summary @vbgl . Something I wonder is about the classification of such commands and their respective attributes. For example, it seems that the runner for an anonymous `Instance` would have pretty different type than the one for a regular definition, wouldn't it?

Zimmi48 commented:
I'm far from an expert of this part of the code. I started looking at it only about two or three weeks ago to see how I could implement my proposed `Lemma :=` feature. I considered the lack of this command variant to be an inconsistency in the way one can declare stuff but I discovered that there were more inconsistencies (see  #6567) and my opinion is that this code is not factorized enough and if it were, the user-level commands would be more consistent. So I'm supportive of your PR at least in principle.

We could try to go further and deprecate `Goal` and have instead `Lemma _ : ...` or `Lemma ? : ...` and `Definition _ := ...` (etc.). (By consistency with the intro-patterns, `_` could mean prove but don't keep (it could be used in particular in `Example _ : ...` and `?` would be keep and name automatically.)

mattam82 commented:
I would be content with an automatically generated name for instances being
independent from the class name and fixed at parsing time or when the
command is added to the document. Easy to fix, simpler design, nobody
should rely on those names anyway.
— Matthieu
Le lun. 29 janv. 2018 à 15:22, Théo Zimmermann <notifications@github.com> a
écrit :

> I'm far from an expert of this part of the code. I started looking at it
> only about two or three weeks ago to see how I could implement my proposed Lemma
> := feature. I considered the lack of this command variant to be an
> inconsistency in the way one can declare stuff but I discovered that there
> were more inconsistencies (see #6567
> <https://github.com/coq/coq/issues/6567>) and my opinion is that this
> code is not factorized enough and if it were, the user-level commands would
> be more consistent. So I'm supportive of your PR at least in principle.
>
> We could try to go further and deprecate Goal and have instead Lemma _ :
> ... or Lemma ? : ... and Definition _ := ... (etc.). (By consistency with
> the intro-patterns, _ could mean prove but don't keep (it could be used
> in particular in Example _ : ... and ? would be keep and name
> automatically.)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/coq/coq/pull/6653#issuecomment-361260113>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGARbPBSVYo7ht2XfAsJ9Ibqb2oS08wks5tPdQegaJpZM4Rsv_A>
> .
>

ejgallego commented:
An interesting feature of `Goal` is that you can do `Save foo.` and have a new lemma named `foo`.

There was some discussion in an older PR about this feature. Personally I like it, but that is true that on the other hand it would make the code better "factorized/parenthesized"

vbgl commented:
The implementation of `Save foo` with anonymous goals seems a little bit fragile: it does not work with anonymous instances, but does work with anything named `Unnamed_thmNNN`, as illustrated by the following snippet.

```
Goal True. exact I. Save foo.

Class Class T := p : T.

Instance : Class nat.
exact 0.
Fail Save bar.
(* This command can only be used for unnamed theorem. *)
Abort.

Instance Unnamed_thm1234 : Class bool.
exact true.
Save bar.
```
---

Back to this PR: I’ve addressed the stylistic comments by Emilio (reuse combinators from the library, do not add dead code and `assert false`).

However, it seems that a better route to follow would be to get rid of anonymous objects and generate fresh names at parsing time. Let me give it a try.

ejgallego commented:
@vbgl +1 . Indeed `Save foo` and related machinery looks a bit clumsy, on the other it is not easy to keep the functionality. Personally not sure still what would be the best way, I think the PR as is now is good, but of course more progress would be great.

gares commented on vernac/lemmas.ml:
```diff
@@ -210,17 +210,17 @@ let save ?export_seff id const uctx do_guard (locality,poly,kind) hook =
+  let avoid = Id.Set.of_list (Proof_global.get_all_proof_names ()) in
+  next_global_ident_away default_thm_id avoid
+
+let check_name_freshness locality ((loc,id), _) =
```
A function called `check_..` that has a result is a bit odd, especially at its use. I think it should be of type ` -> unit` and be used as
```
check_name_freshness id;
... id ...
```
Alternatively call it something like `get_id_if_fresh`

vbgl commented:
Just addressed Enrico’s remark (`check… id` now has type unit).

Regarding (fresh) name generation at parsing time, this seems rather difficult: nothing prevents other commands (e.g., plug-ins) to generate colliding names after parsing of the whole document.
